watches.go

```diff
@@ -52,39 +66,71 @@ func (wm *watchManager) handleWatchRequest(w http.ResponseWriter, r *http.Reques
 		http.Error(w, fmt.Sprintf("invalid 'fromIndex': %v", err), http.StatusBadRequest)
 		return
 	}
-
-	go wm.manageWatch(conn, topic, genID, idx)
+	log.Printf("Connection accepted from %v", r.Host)
+	if err = wm.manageWatch(w, topic, genID, idx); err != nil {
+		log.Printf("Error: watch %v\n", err)
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
 }
 
-func (wm *watchManager) manageWatch(conn *websocket.Conn, topic, genID string, idx uint64) {
-	log.Printf("Connection accepted from %v", conn.RemoteAddr())
-	defer closeConn(conn)
+func (wm *watchManager) manageWatch(w http.ResponseWriter, topic string, genID string, idx uint64) error {
+	cn, canNotifyClose := w.(http.CloseNotifier)
+	flusher, canFlush := w.(http.Flusher)
+	if !canNotifyClose || !canFlush {
+		return errors.New("Error: cannot stream")
+	}
+	aw := newActiveWatch(wm, topic, genID, idx, httputil.NewChunkedWriter(w), flusher)
 	for {
-		msgsResponse, err := wm.store.get(topic, genID, idx)
-		if err != nil {
-			handleError(err, conn)
-			return
-		}
-		if msgsLength := len(msgsResponse.Messages); msgsLength > 0 {
-			if err := conn.WriteJSON(msgsResponse); err != nil {
-				handleError(err, conn)
-				return
+		select {
+		case <-cn.CloseNotify():
+			aw.close()
+			return nil
+		default:
+			if err := aw.handleNewMessages(); err != nil {
+				return err
 			}
-			genID = msgsResponse.GenerationID
-			idx = msgsResponse.Messages[msgsLength-1].Index + 1
 		}
 		time.Sleep(wm.pushInterval)
 	}
 }
 
-func closeConn(conn *websocket.Conn) {
-	log.Printf("Terminating connection to %v", conn.RemoteAddr())
-	if err := conn.Close(); err != nil {
-		log.Printf("[WARNING] error closing connection: %v", err)
+func (aw *activeWatch) handleNewMessages() error {
+	var err error
+	var msgsResponse *MessagesResponse
+	if msgsResponse, err = aw.newMessages(); err != nil {
+		return err
+	}
+	if msgsLength := len(msgsResponse.Messages); msgsLength > 0 {
+		if err := aw.writeChunk(msgsResponse); err != nil {
+			return err
+		}
+		aw.genID = msgsResponse.GenerationID
+		aw.idx = msgsResponse.Messages[msgsLength-1].Index + 1
 	}
+	return nil
+}
+
+func (aw *activeWatch) newMessages() (*MessagesResponse, error) {
+	return aw.wm.store.get(aw.topic, aw.genID, aw.idx)
 }
 
-func handleError(err error, conn *websocket.Conn) error {
-	log.Printf("Closing connection due to error: %v", err)
-	return conn.WriteMessage(websocket.CloseMessage, websocket.FormatCloseMessage(websocket.CloseInternalServerErr, err.Error()))
+func (aw *activeWatch) writeChunk(msgs *MessagesResponse) error {
+	var marshalled []byte
+	var err error
```

Comment:
I had to move the assignment outside of the if (since marshalled is needed later)